### ISSUE TITLE
[Federation] @key agreement and usage clarification

### ISSUE THREAD

**Issue by lennyburdette** [apollographql/apollo-server#3202]
This is a very subtle issue (or possibly a misunderstanding)! Please take a look at this reproduction repo for more details: https://github.com/lennyburdette/apollo-federation-key-test

**Should there be a requirement that `@key` directives have to agree?**

The gateway doesn't enforce agreement today. In my example, the services have different indexes and lookup requirements so disagreement in keys makes sense.

**When a services declares the `@key` for an entity, does that specify its _own_ requirements for lookup by representation, or does it declare what values it can _provide_ for lookups in other services?**

It seems like a possible outcome is the need for additional schema validation: if a service declares a compound key like `@key(fields: "id business { id }`), all other services must be able to fulfill that key (they must have an `id` field and a `business` field.)

**Comment by JacksonKearl** [comment]
Hey @lennyburdette,

First off, your situation sounds like a perfect use case for [multiple primary keys](https://www.apollographql.com/docs/apollo-server/federation/advanced-features/#multiple-primary-keys). Which would allow you to declare that a `Payment` can be identified by either an `id` OR an `id business { id }`:
```gql
type Payment @key(fields: "id") @key(fields: "id business { id }") {
   ...
}
```

If you wrote it like that, and used a type generator for federation, you'd see:
```ts
/** @type import('./resolver-types').Resolvers */
const resolvers = {
  Payment: {
    __resolveReference(ref) {
      // ERR: Property 'business' does not exist on type '{ id: string; } | { id: string; business: { id: string; }; }' 
      return (DB[ref.business.id] || {})[ref.id]; 
    }
  },
  Query: {
    payment(_, { id, businessID }) {
      return (DB[businessID] || {})[id];
    }
  }
};
```
(Which is exactly the problem you encountered)


But, more to your point, yes I believe there should probably be a validation step that ensures your specified key selection set is a "real" one. If @abernix is on the same page, I will add it.

**Comment by lennyburdette** [comment]
Thanks for the fast response @JacksonKearl ! The validation PR make a lot of sense. 

I'm not entirely sure how I can use multiple keys here, because I have one service (A) that _requires_ the compound key. So the other service (B) must agree so that the gateway fetches `business { id }` from B in order to request entities from A. I can just standardize on the compound key and make sure that all services can provide that data. The validation will help a lot there.

But I had forgotten that multiple `@key`s was possible, so thanks for the reminder! 😃

**Comment by trevor-scheer** [comment]
Closed via #4498
